pkg/reconcilermanager/controllers/reconciler_base.go

```diff
@@ -236,17 +263,75 @@ func (r *reconcilerBase) createOrPatchDeployment(ctx context.Context, declared *
 		r.log.Error(err, "Managed object update failed, deleting and re-creating",
 			logFieldObject, dRef.String(),
 			logFieldKind, kind)
-		if err := r.client.Delete(ctx, declared); err != nil {
+		if err := deploymentResourceInterface.Delete(ctx, dRef.Name, metav1.DeleteOptions{}); err != nil {
 			return controllerutil.OperationResultNone, err
 		}
-		if err := r.client.Create(ctx, declared); err != nil {
+		data, err := json.Marshal(declared)
+		if err != nil {
+			return controllerutil.OperationResultNone, fmt.Errorf("failed to marshal declared deployment object to byte array: %w", err)
+		}
+		_, err = deploymentResourceInterface.Patch(ctx, dRef.Name, types.ApplyPatchType, data, metav1.PatchOptions{FieldManager: reconcilermanager.ManagerName, Force: &forcePatch})
+		if err != nil {
 			return controllerutil.OperationResultNone, err
 		}
 	}
-
+	if deploymentAfterPatch.GetGeneration() == unObjCurrent.GetGeneration() {
+		return controllerutil.OperationResultNone, nil
+	}
 	return controllerutil.OperationResultUpdated, nil
 }
 
+// preprocessDeployment delete all the fields in allowlist from unstructured object and convert the unstructured object to Deployment object
+func preprocessDeployment(allowList []string, unstructuredDeployment *unstructured.Unstructured) (*appsv1.Deployment, error) {
+	for _, path := range allowList {
+		if err := deleteFields(unstructuredDeployment.Object, path); err != nil {
+			return nil, err
+		}
+	}
+	var resultDeployment appsv1.Deployment
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredDeployment.Object, &resultDeployment); err != nil {
+		return nil, fmt.Errorf("failed to convert from current reconciler unstructured object to deployment object: %w", err)
+	}
+	return &resultDeployment, nil
+}
+
+type deploymentProcessResult struct {
+	isDeploymentSame     bool
```

Comment:
Are these two booleans just inverse of each other? What's the difference?

Reply:
isDeploymentAdjusted is using to detect if Autopilot make any adjustment to the resources requirement we defined.  Resources adjusted by Autopilot doesn't mean the declared and current deployments are not same. We can tell that if Autopilot did adjustment based on what we declared, if yes, we still consider the current deployment is same with declared deployment and no more PATCH needed. So the isDeployment same is used to decide if a PATCH is needed, the isDeploymentAdjusted is only used to notice the user that the resource defined by us or by them was adjusted by Autopilot.